src/pages/experiments/[experimentId]/data-exploration/components/embedding/Embedding.jsx

```diff
@@ -37,75 +37,86 @@ const Embedding = (props) => {
   const [selectedIds, setSelectedIds] = useState([]);
   const [cellColors, setCellColors] = useState({});
-  const currentView = useRef(focusedGene ? 'expression' : 'cellSet');
+
   const [cellInfoVisible, setCellInfoVisible] = useState(true);
 
+  // Load the embedding if it isn't already.
   useEffect(() => {
     if (!data && isBrowser) {
       dispatch(loadEmbedding(experimentId, embeddingType));
     }
   }, []);
 
-  const getCellColors = (coloringMode) => {
-    if (coloringMode === 'expression') {
-      return colorByGeneExpression(focusedExpression);
-    }
+  // Handle focus change (e.g. a cell set or gene or metadata got selected).
+  // Also handle here when the cell set properties or hierarchy change.
+  useEffect(() => {
+    const { store, key } = focusData;
 
-    if (coloringMode === 'cellSet') {
-      const colors = renderCellSetColors(cellSetSelected, cellSetProperties);
-      return colors;
-    }
+    switch (store) {
+      // For genes/continous data, we cannot do this in one go,
+      // we need to wait for the thing to load in first.
+      case 'genes': {
+        dispatch(loadGeneExpression(experimentId, [key]));
+        return;
+      }
 
-    return {};
-  };
+      // Cell sets are easy, just return the appropriate color and set them up.
+      case 'cellSet': {
+        if (isBrowser) setCellColors(returnCellColoring(focusData));
+        break;
+      }
 
-  // Handle a cell set being selected for coloring.
-  useEffect(() => {
-    currentView.current = 'cellSet';
-    if (!loadingColors && isBrowser) setCellColors(getCellColors('cellSet'));
-  }, [cellSetSelected]);
-
-  // Handle focusing/defocusing of genes. This will set a loading
-  // state and try to fetch the expression level from the store or the backend.
-  // This MUST be after the hook of cellSetSelected
-  // to be preferentially loaded over cell set selections.
-  useEffect(() => {
-    // If we defocused a gene, set it back to cell set coloring view.
-    if (!focusedGene) {
-      currentView.current = 'cellSet';
-      setCellColors(getCellColors('cellSet'));
-      return;
+      // If there is no focus, we can just delete all the colors.
+      default: {
+        setCellColors({});
+        break;
+      }
     }
+  }, [focusData, cellSetHierarchy, cellSetProperties]);
 
-    currentView.current = 'expression';
-    if (isBrowser) dispatch(loadGeneExpression(experimentId, [focusedGene]));
-  }, [focusedGene]);
-
-  // Handle loading of expression for focused gene.
+  // // Handle loading of expression for focused gene.
   useEffect(() => {
     if (!focusedExpression) {
       return;
     }
-    setCellColors(getCellColors('expression'));
+
+    setCellColors(returnCellColoring(focusData));
   }, [focusedExpression]);
 
+  const returnCellColoring = (focusInfo) => {
+    const { store, key } = focusInfo;
+
+    if (store === 'genes') {
+      return colorByGeneExpression(focusedExpression);
+    }
```

Comment:
this if is useless - the function `returnCellColoring` only gets called when `focusStore = 'cellSet'`